### ISSUE TITLE
False positive for `unnecessary_parenthesis`

### ISSUE THREAD

**Issue by Pante** [dart-lang/linter#2944]
* Dart SDK Version: 2.14

I currently have the following snippet of code.
```dart
void foo() {
  final items = [];
  (() => [].add('something')).compose(() {});
}


extension A on void Function() {

  void Function() compose(void Function() other) => () {
    this();
    other();
  };

}
```

`dart analyze` falsely reports that the parenthesis around `(() => [].add('something'))` can be removed. However, it cannot be removed since the code's semantics will change and become `() => [].add('something').compose(() {})`. 

I think it's also worth noting that the parenthesis can be removed if the function used `{ ... }` instead of `=>`.
